dev/publish_bom.py

```diff
@@ -0,0 +1,186 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import datetime
+import os
+import socket
+import sys
+import yaml
+
+from github import Github
+from github.Gist import Gist
+from github.InputFileContent import InputFileContent
+
+from generate_bom import BomGenerator
+from spinnaker.run import check_run_quick, run_quick
+
+SERVICES = 'services'
+VERSION = 'version'
+
+COMPONENTS = [
+  'clouddriver',
+  'deck',
+  'echo',
+  'front50',
+  'gate',
+  'igor',
+  'orca',
+  'rosco',
+  'fiat',
+  'spinnaker-monitoring',
+  'spinnaker'
+]
+
+class BomPublisher(BomGenerator):
+
+  def __init__(self, options):
+    self.__rc_version = options.rc_version
+    self.__bom_dict = {}
+    self.__release_version = options.release_version
+    self.__github_publisher = options.github_publisher
+    self.__changelog_file = options.changelog_file
+    self.__github_token = options.github_token
+    self.__gist_user = options.gist_user
+    self.__github = Github(self.__gist_user, self.__github_token)
+    self.__patch_release = options.patch_release
+    self.__alias = options.bom_alias # Flag inherited from BomGenerator.
+    super(BomPublisher, self).__init__(options)
+
+  def unpack_bom(self):
+    """Load the release candidate BOM into memory.
+    """
+    bom_yaml_string = run_quick('hal versions bom {0} --color false'
+                                .format(self.__rc_version), echo=False).stdout.strip()
+    print bom_yaml_string
+    self.__bom_dict = yaml.load(bom_yaml_string)
+    print self.__bom_dict
+
+  def publish_release_bom(self):
+    """Read, update, and publish a release candidate BOM.
+    """
+    new_bom_file = '{0}.yml'.format(self.__release_version)
+    self.__bom_dict[VERSION] = self.__release_version
+    self.write_bom_file(new_bom_file, self.__bom_dict)
+    self.publish_bom(new_bom_file)
+    # Re-write the 'latest' Spinnaker version.
+    # TODO(jacobkiefer): Update 'available versions' with Halyard when that feature is ready.
+    if self.__alias:
+      alias_file = '{0}.yml'.format(self.__alias)
+      self.write_bom_file(alias_file, self.__bom_dict)
+      self.publish_bom(alias_file)
+
+  def publish_changelog_gist(self):
+    """Publish the changelog as a github gist.
+    """
+    description = 'Changelog for Spinnaker {0}'.format(self.__release_version)
+    with open(self.__changelog_file, 'r') as clog:
+      raw_content_lines = clog.readlines()
+      spinnaker_version = '# Spinnaker {0}\n'.format(self.__release_version)
+      # Re-write the correct Spinnaker version at the top of the changelog.
+      # Also add some identifying information.
+      raw_content_lines[0] = spinnaker_version
+      timestamp = '{:%Y-%m-%d %H:%M:%S}'.format(datetime.datetime.now())
```

Comment:
Check out 'datetime' in [this article](https://pyformat.info/). There are a bunch of ways to use `format` essentially.